core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -0,0 +1,643 @@
+
+	if (size_todo) {
+		desc = hashdata->descriptor;
+		caam_desc_init(desc);
+		caam_desc_add_word(desc, DESC_HEADER(0));
+
+		/* There are blocks to hash - Create the Descriptor */
+		if (hashdata->ctx.length) {
+			HASH_TRACE("Update Operation");
+			/* Algo Operation - Update */
+			caam_desc_add_word(desc, HASH_UPDATE(alg->type));
+			/* Running context to restore */
+			caam_desc_add_word(desc,
+					   LD_NOIMM(CLASS_2, REG_CTX,
+						    hashdata->ctx.length));
+			caam_desc_add_ptr(desc, hashdata->ctx.paddr);
+		} else {
+			HASH_TRACE("Init Operation");
+
+			/* Algo Operation - Init */
+			caam_desc_add_word(desc, HASH_INIT(alg->type));
+
+			hashdata->ctx.length = alg->size_ctx;
+		}
+
+		if (hashdata->blockbuf.filled != 0) {
+			/* Add the temporary buffer */
+			caam_desc_add_word(desc,
+					   FIFO_LD_EXT(CLASS_2, MSG, NOACTION));
+			caam_desc_add_ptr(desc, hashdata->blockbuf.buf.paddr);
+			caam_desc_add_word(desc, hashdata->blockbuf.filled);
+
+			/* Clean the circular buffer data to be loaded */
+			cache_operation(TEE_CACHECLEAN,
+					hashdata->blockbuf.buf.data,
+					hashdata->blockbuf.filled);
+			hashdata->blockbuf.filled = 0;
+		}
+
+		/* Add the input data multiple of blocksize */
+		caam_desc_add_word(desc, FIFO_LD_EXT(CLASS_2, MSG, LAST_C2));
+		caam_desc_add_ptr(desc, paddr_data);
+		caam_desc_add_word(desc, size_inmade);
+
+		/* Clean the input data to be loaded */
+		cache_operation(TEE_CACHECLEAN, (void *)data, size_inmade);
+
+		/* Save the running context */
+		caam_desc_add_word(desc, ST_NOIMM(CLASS_2, REG_CTX,
+						  hashdata->ctx.length));
+		caam_desc_add_ptr(desc, hashdata->ctx.paddr);
+
+		HASH_DUMPDESC(desc);
+
+		/* Invalidate Context register */
+		cache_operation(TEE_CACHEINVALIDATE, hashdata->ctx.data,
+				hashdata->ctx.length);
+
+		jobctx.desc = desc;
+		retstatus = caam_jr_enqueue(&jobctx, NULL);
+
+		if (retstatus == CAAM_NO_ERROR) {
+			ret = TEE_SUCCESS;
+			HASH_DUMPBUF("CTX", hashdata->ctx.data,
+				     hashdata->ctx.length);
+		} else {
+			HASH_TRACE("CAAM Status 0x%08" PRIx32, jobctx.status);
+			ret = job_status_to_tee_result(jobctx.status);
+		}
+	} else {
+		ret = TEE_SUCCESS;
+
+		if (size_topost) {
+			/* All input data must be saved */
+			size_inmade = 0;
+		}
+	}
+
+	if ((size_topost) && (data)) {
+		struct caambuf indata = { .data = (uint8_t *)data,
+					  .length = inlen };
+
+		HASH_TRACE("Post %zu of input len %zu made %zu", size_topost,
+			   len, size_inmade);
+		ret = caam_cpy_block_src(&hashdata->blockbuf, &indata,
+					 size_inmade);
+	}
+
+exit_update:
+	if (ret != TEE_SUCCESS)
+		do_free_intern(hashdata);
+
+	return ret;
+}
+
+/*
+ * Finalize the Hash operation
+ *
+ * @ctx     Operation Software context
+ * @len     Digest buffer length
+ * @digest  [out] Hash digest buffer
+ */
+static TEE_Result do_final(struct crypto_hash_ctx *ctx, uint8_t *digest,
+			   size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+
```

Comment:
Please drop the empty lines in the variables declaration

Reply:
ok